### ISSUE TITLE
Triangle().area for zero-area triangle results in AttributeError

### ISSUE THREAD

**Issue by tylerharvey** [sympy/sympy#20460]
`Triangle(Point(0,0),Point(0,1),Point(0,0.5)).area`
results in 
`---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-11-4686e8494747> in <module>
----> 1 Triangle(Point(0,0),Point(0,1),Point(0,0.5)).area

AttributeError: 'Segment2D' object has no attribute 'area'`

I'm new to the project and understand that there may be some good reasons to cast a zero-area Triangle to a Segment2D; I would also strongly argue that Triangle() and its functions should work correctly for arbitrary user-inputted or randomly-generated points. Perhaps a simple fix is to implement Segment2D.area and any other functions for finite-area polygons?

**Comment by oscarbenjamin** [comment]
I guess if a 2D shape like a triangle will evaluate to something like a segment then the segment needs to have the same methods and properties so adding area would make sense.

**Comment by KatsuragiCSL** [comment]
A fixes is done by https://github.com/sympy/sympy/pull/20469, simply added `area` attribute to `Segment2D` class

**Comment by sidhu1012** [comment]
> AttributeError: 'Segment2D' object has no attribute 'area'

`Triangle(Point(0,0),Point(0,1),Point(0,0.5))` has all it's three points on same line. As much as I know, lines or line segments as in this case doesn't have area.

> I'm new to the project and understand that there may be some good reasons to cast a zero-area Triangle to a Segment2D; I would also strongly argue that Triangle() and its functions should work correctly for arbitrary user-inputted or randomly-generated points

A triangle is never formed in case of points on a straight line, as in this case all points lie on y-axis, So as I said it's better to give error as the Triangle is never formed.


If points don't lie on same line `Triangle.area` works fine.

**Comment by tylerharvey** [comment]
> 
> 
> A fixes is done by #20469, simply added `area` attribute to `Segment2D` class

Great!

**Comment by sylee957** [comment]
It may not be a good choice to degenerate the objects automatically, or area needed to be added for `Point` too.

**Comment by KatsuragiCSL** [comment]
I think adding area attribute to Point will be good.

**Comment by sidhu1012** [comment]
> It may not be a good choice to degenerate the objects automatically, or area needed to be added for `Point` too.

Yes agreed!

**Comment by sylee957** [comment]
The only question I would raise is that we have to design attributes consistently.
Once area is added to some of the 0D or 1D geometry objects, it breaks some some consistency that area 
And there are also future possibilities that we need 3D volume, or any n-dimensional volumes, so adding area can also logically conclude that n-dimensional volume should be added for all the lesser dimensional objects.
So beside segment, point, we also need to conclude how to comprehend area for line, curve, 3d geometries, ...
I'd expect to see some mathematical consistency behind any idea.

**Comment by tylerharvey** [comment]
Can anyone comment on the design choice behind casting a zero-area Triangle to a Segment2d? I wouldn't particularly expect to get, say, length out of a zero-area triangle. I can't see how someone intending to make a triangle would appreciate getting line segment attributes and methods instead.

**Comment by sidhu1012** [comment]
> Can anyone comment on the design choice behind casting a zero-area Triangle to a Segment2d? I wouldn't particularly expect to get, say, length out of a zero-area triangle. I can't see how someone intending to make a triangle would appreciate getting line segment attributes and methods instead.

I think the idea behind this could be that in case of zero area traingle all points lie on a same line forming Segment rather than a triangle.

**Comment by tylerharvey** [comment]
> 
> 
> > Can anyone comment on the design choice behind casting a zero-area Triangle to a Segment2d? I wouldn't particularly expect to get, say, length out of a zero-area triangle. I can't see how someone intending to make a triangle would appreciate getting line segment attributes and methods instead.
> 
> I think the idea behind this could be that in case of zero area traingle all points lie on a same line forming Segment rather than a triangle.

Yes, and a zero-area triangle is still a triangle. Calling it one or the other has usage implications and I don't see a usage argument for calling is a line segment.

**Comment by sylee957** [comment]
I'm not sure if the geometry module was actively maintained, but I think that most of the original code were designed to switch between objects when they were degenerate.
So we should do consistent change, for other objects like circle, polygon, ...
If you are using for downstream projects, the only immediate option can be checking the types of the objects there.

**Comment by almighty123789** [comment]
@sylee957  Would adding an area variable equal to zero to Point and LinearEntity(the base class for Segment 2D, Line etc) do the job?
I too think that these should have an area attribute which is assigned 0.

**Comment by sidhu1012** [comment]
If area attribute is going to be added to lined it would be better to add area attribute to class LinearEntity and all other linear entity classes . It would make changes symmetric and would give area = 0 to all lines.

**Comment by almighty123789** [comment]
@sylee957 I have added an area attribute to the LinearEntity and Point Class, along with the relevant tests. This potentially resolves this issue.

**Comment by sylee957** [comment]
Unfortunately, I think that we need more commentary about how to define length, area, volumes over more wider variety of objects in more mathematically rigorous sense. Or I think that we may end up having a temporal change that is difficult to resolve in the future.

**Comment by oscarbenjamin** [comment]
Defining area for a line seems straight-forward to me. The geometry module doesn't handle anything more than 3D.

**Comment by almighty123789** [comment]
I agree with what @oscarbenjamin  has said. This change shouldn't really conflict with any changes that geometry module might undergo, since as the above issue shows, we would need an area attribute specifically for lines and points to deal with degenerate **polygons.**
This can be seen as 
_Example taken from sympy live_
`>>> v=Polygon(Point(0,0),Point(0,0),Point(0,0),Point(0,0))`
`>>> v.area`
`Traceback (most recent call last):`
  `File "<string>", line 1, in <module>`
`AttributeError: 'Point2D' object has no attribute 'area'` 

These changes once done, would potentially resolve these types of issues too, i.e where degenerate polygons are involved.

**Comment by sylee957** [comment]
There is difference a that line is a parametrized from 1 dimensional region and surface is parametrized from 2 dimensional region, regardless of the dimension of the coordinates, and area is usually only defined over the latter geometries.  (So 1d line, 2d line, 3d line, 2d circle, 3d circle, ... can only have length, 2d triangle, 3d triangle, 3d rectangle, 3d sphere, ..., can only have area and only box and ball can have volume)
We can have the adhoc statement that line is defined over ![image](https://user-images.githubusercontent.com/34944973/104984245-b32f3b80-5a51-11eb-9d9a-28a2b09043df.png) rather than ![image](https://user-images.githubusercontent.com/34944973/104984279-c4784800-5a51-11eb-9d79-06be4ec2e08a.png) such that the area is zero

**Comment by smichr** [comment]
Degeneracy is something one should expect when working with SymPy: `Eq(1, 1).lhs` will give an attribute error because the equality reduces to a boolean. I don't think it is reasonable to give LinearEntity and Point all the attributes of Triangles; and if not for other attributes, then not for area. It might make more sense to retain all points in a degenerate Triangle and only resolve to the degenerate case with simplification. But that will create other problems, I imagine, when trying to compute certain quantities like the various centers.

But here is a case where the degeneracy is not the same (and this is not right):
```python
>>> Triangle((0,0),(1,0),(-1,0))  # sss=(1,2,1)
Point2D(-1, 0)
>>> Triangle((-1,0),(1,0),(0,0))  # sss=(2,1,1)
Segment2D(Point2D(-1, 0), Point2D(0, 0))
```
